File: backend/internal/types/network.go
package types

// ClientMessage represents messages sent from the frontend to the server.
type ClientMessage struct {
	Type     string `json:"type"`
	Col      *int   `json:"col,omitempty"`
	GameID   string `json:"gameId,omitempty"`
	Username string `json:"username,omitempty"`
}

// ServerMessage mirrors the frontend contract.
type ServerMessage struct {
	Type        string  `json:"type"`
	GameID      string  `json:"gameId,omitempty"`
	You         int     `json:"you,omitempty"`
	Opponent    string  `json:"opponent,omitempty"`
	Board       [][]int `json:"board,omitempty"`
	CurrentTurn int     `json:"currentTurn,omitempty"`
	Result      string  `json:"result,omitempty"`
	Message     string  `json:"message,omitempty"`
}

File: src/types/network.ts
export type ClientMessage =
  | { type: 'MAKE_MOVE'; col: number; gameId: string }
  | { type: 'RECONNECT'; username: string; gameId?: string };

export type ServerMessage =
  | { type: 'GAME_START'; gameId: string; you: 1 | 2; opponent: string }
  | { type: 'BOARD_UPDATE'; board: number[][]; currentTurn: 1 | 2 }
  | { type: 'GAME_OVER'; result: 'WIN' | 'LOSS' | 'DRAW'; board: number[][] }
  | { type: 'INFO'; message: string };
